File: helixtrajectoryj/src/main/java/org/team2363/helixtrajectory/SwerveDrivetrain.java
package org.team2363.helixtrajectory;

import static org.team2363.util.ObjectChecker.requireNonNullAndWrapUnmodifiable;

import java.util.List;

public class SwerveDrivetrain extends HolonomicDrivetrain {

    public final List<? extends SwerveModule> modules;

    public SwerveDrivetrain(double mass, double momentOfInertia, List<? extends SwerveModule> modules, Obstacle bumpers) throws NullPointerException {
        super(mass, momentOfInertia, bumpers);

        this.modules = requireNonNullAndWrapUnmodifiable(modules, "List of swerve modules cannot be null", "Swerve module cannot be null");
    }

    @Override
    public String toString() {
        return "{\"mass\": " + mass
            + ", \"moment_of_inertia\": " + momentOfInertia
            + ", \"modules\": " + modules
            + ", \"bumpers\": " + bumpers
            + "}";
    }
}

File: helixtrajectorycpp/src/cpp/SwerveDrivetrain.cpp
#include "drivetrain/SwerveDrivetrain.h"

#include <iostream>
#include <cmath>
#include <vector>

#include <casadi/casadi.hpp>

#include "drivetrain/HolonomicDrivetrain.h"
#include "drivetrain/SwerveDrivetrain.h"
#include "obstacle/Obstacle.h"

namespace helixtrajectory {

    SwerveDrivetrain::SwerveDrivetrain(double mass, double momentOfInertia, const std::vector<SwerveModule>& modules) :
        HolonomicDrivetrain(mass, momentOfInertia), modules(modules) {
    }

    std::ostream& operator<<(std::ostream& stream, const SwerveDrivetrain& swerveDrivetrain) {
        return stream << "{\"mass\": " << swerveDrivetrain.mass
               << ", \"moment_of_inertia\": " << swerveDrivetrain.momentOfInertia
               << ", \"modules\": " << swerveDrivetrain.modules
               << "}";
    }
}